### PR TITLE
fix: bring back delayed leaf detection in listing

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: bring back delayed leaf detection in listing

## Motivation and Context
improve List performance

## How to test this PR?
No functionality should change

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [x] Unit tests needed

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#d227636b-4d0c-44cc-bc53-16edd44ad73d) |
| mint-zoned.sh | :x: [more...](#eea92f6a-f095-4485-a634-c3faff197590) |
| mint-gateway-nas.sh | :x: [more...](#2d7d3867-480e-4660-a2cc-936ff258c93c) |
<a name="d227636b-4d0c-44cc-bc53-16edd44ad73d"></a>
### 10346-5c7da30/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:32632
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 97b9cd4d62fb:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 5 minutes and 35 seconds
(4/15) Running aws-sdk-ruby tests ... done in 19 seconds
(5/15) Running awscli tests ... done in 2 minutes and 36 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 26 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 31 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 40 seconds
{
  "args": {
    "bucketName": "minio-go-test-b94jua1cjcl0bwck",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 644,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 6 minutes and 14 seconds
(10/15) Running minio-js tests ... FAILED in 47 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 81,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:678:10) at endReadableNT (_stream_readable.js:1244:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... done in 10 minutes and 16 seconds
(11/15) Running s3cmd tests ... done in 1 minutes and 20 seconds
(12/15) Running s3select tests ... done in 32 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="2d7d3867-480e-4660-a2cc-936ff258c93c"></a>
### 10346-5c7da30/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:30377
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp cd9c1718014a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 1 minutes and 58 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 36 seconds
(8/15) Running minio-dotnet tests ... done in 29 seconds
(9/15) Running minio-go tests ... done in 1 minutes and 0 seconds
(10/15) Running minio-java tests ... done in 22 seconds
(11/15) Running minio-js tests ... done in 39 seconds
(12/15) Running minio-py tests ... FAILED in 1 minutes and 16 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-1d1913d2-86e0-467f-832d-53afd3e7d6c3",
    "object_name": "a1d10f11-7f39-4551-9b6d-e8aee739afbe"
  },
  "message": "(\"Connection broken: ConnectionResetError(104, 'Connection reset by peer')\", ConnectionResetError(104, 'Connection reset by peer'))",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/response.py\", line 436, in _error_catcher\n    yield\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/response.py\", line 518, in read\n    data = self._fp.read(amt) if not fp_closed else b\"\"\n  File \"/usr/lib/python3.6/http/client.py\", line 459, in read\n    n = self.readinto(b)\n  File \"/usr/lib/python3.6/http/client.py\", line 503, in readinto\n    n = self.fp.readinto(b)\n  File \"/usr/lib/python3.6/socket.py\", line 586, in readinto\n    return self._sock.recv_into(b)\nConnectionResetError: [Errno 104] Connection reset by peer\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 147, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1671, in test_thread_safe\n    raise exceptions[0]\n  File \"/mint/run/core/minio-py/tests.py\", line 1631, in get_object_and_check\n    _CLIENT.fget_object(bucket_name, object_name, local_file)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 817, in fget_object\n    for data in response.stream(amt=1024 * 1024):\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/response.py\", line 575, in stream\n    data = self.read(amt=amt, decode_content=decode_content)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/response.py\", line 540, in read\n    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)\n  File \"/usr/lib/python3.6/contextlib.py\", line 99, in __exit__\n    self.gen.throw(type, value, traceback)\n  File \"/usr/local/lib/python3.6/dist-packages/urllib3/response.py\", line 454, in _error_catcher\n    raise ProtocolError(\"Connection broken: %r\" % e, e)\nurllib3.exceptions.ProtocolError: (\"Connection broken: ConnectionResetError(104, 'Connection reset by peer')\", ConnectionResetError(104, 'Connection reset by peer'))\n",
  "duration": 1688
}
(12/15) Running s3cmd tests ... done in 16 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="eea92f6a-f095-4485-a634-c3faff197590"></a>
### 10346-5c7da30/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31730
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 2fdb8d074bf5:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... FAILED in 2 minutes and 1 seconds
{
  "alert": "",
  "args": {
    "bucketName": "aws-sdk-go-test-4dys3wzqf3j9xq",
    "expiry": 60000000000,
    "objectName": "presignedTest"
  },
  "duration": 121195,
  "error": "BucketAlreadyOwnedByYou: Your previous request to create the named bucket succeeded and you already own it.\n\tstatus code: 409, request id: 162E7A4B76E2BF09, host id: ",
  "function": "PresignedPut",
  "message": "AWS SDK Go CreateBucket Failed",
  "name": "aws-sdk-go",
  "status": "FAIL"
}
(1/15) Running aws-sdk-java tests ... done in 1 seconds
(2/15) Running aws-sdk-php tests ... done in 44 seconds
(3/15) Running aws-sdk-ruby tests ... done in 6 seconds
(4/15) Running awscli tests ... done in 2 minutes and 18 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 48 seconds
(7/15) Running minio-dotnet tests ... done in 42 seconds
(8/15) Running minio-go tests ... done in 1 minutes and 52 seconds
(9/15) Running minio-java tests ... done in 1 minutes and 10 seconds
(10/15) Running minio-js tests ... done in 1 minutes and 2 seconds
(11/15) Running minio-py tests ... done in 3 minutes and 9 seconds
(12/15) Running s3cmd tests ... done in 22 seconds
(13/15) Running s3select tests ... done in 10 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10346-5c7da30

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#c58ad248-42a6-4233-9c40-e0341845a9ed) |
| mint-erasure.sh | :x: [more...](#c9c712b2-3b27-4347-abdd-26039e8dd0e6) |
<a name="c58ad248-42a6-4233-9c40-e0341845a9ed"></a>
### 10346-8f9c04a/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:30536
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 97fbb14b3482:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 5 minutes and 35 seconds
(4/15) Running aws-sdk-ruby tests ... done in 21 seconds
(5/15) Running awscli tests ... done in 2 minutes and 48 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 24 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 33 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 38 seconds
{
  "args": {
    "bucketName": "minio-go-test-izirg1tfu9hp565m",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 593,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 6 minutes and 11 seconds
(10/15) Running minio-js tests ... FAILED in 46 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 75,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:678:10) at endReadableNT (_stream_readable.js:1244:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... done in 11 minutes and 13 seconds
(11/15) Running s3cmd tests ... done in 1 minutes and 31 seconds
(12/15) Running s3select tests ... done in 33 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="c9c712b2-3b27-4347-abdd-26039e8dd0e6"></a>
### 10346-8f9c04a/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:31783
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp fa84c10b2cfd:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... FAILED in 2 minutes and 5 seconds
{
  "name": "awscli",
  "duration": 68546,
  "function": "aws --endpoint-url http://minio-dev6.minio.io:31783 s3 sync --no-progress /mint/data s3://awscli-mint-test-bucket-13616/\n",
  "status": "FAIL",
  "error": "upload: ../../../data/datafile-0-b to s3://awscli-mint-test-bucket-13616/datafile-0-bupload: ../../../data/datafile-1-kB to s3://awscli-mint-test-bucket-13616/datafile-1-kBupload: ../../../data/datafile-10-kB to s3://awscli-mint-test-bucket-13616/datafile-10-kBupload: ../../../data/datafile-1-b to s3://awscli-mint-test-bucket-13616/datafile-1-bupload: ../../../data/datafile-33-kB to s3://awscli-mint-test-bucket-13616/datafile-33-kBupload: ../../../data/datafile-100-kB to s3://awscli-mint-test-bucket-13616/datafile-100-kBupload: ../../../data/datafile-1-MB to s3://awscli-mint-test-bucket-13616/datafile-1-MBupload: ../../../data/datafile-1.03-MB to s3://awscli-mint-test-bucket-13616/datafile-1.03-MBupload: ../../../data/datafile-5-MB to s3://awscli-mint-test-bucket-13616/datafile-5-MBupload: ../../../data/datafile-5243880-b to s3://awscli-mint-test-bucket-13616/datafile-5243880-bupload: ../../../data/datafile-6-MB to s3://awscli-mint-test-bucket-13616/datafile-6-MBupload: ../../../data/datafile-10-MB to s3://awscli-mint-test-bucket-13616/datafile-10-MBupload: ../../../data/datafile-11-MB to s3://awscli-mint-test-bucket-13616/datafile-11-MBupload: ../../../data/datafile-65-MB to s3://awscli-mint-test-bucket-13616/datafile-65-MBupload failed: ../../../data/datafile-129-MB to s3://awscli-mint-test-bucket-13616/datafile-129-MB An error occurred (NoSuchBucket) when calling the UploadPart operation: The specified bucket does not exist"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 34 seconds
(7/15) Running minio-dotnet tests ... done in 29 seconds
(8/15) Running minio-go tests ... done in 1 minutes and 18 seconds
(9/15) Running minio-java tests ... done in 18 seconds
(10/15) Running minio-js tests ... done in 51 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 49 seconds
(12/15) Running s3cmd tests ... done in 14 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10346-8f9c04a

File: cmd/tree-walk.go [hunk 1]
```diff
@@ -51,15 +51,80 @@ func filterMatchingPrefix(entries []string, prefixEntry string) []string {
...
+}
+
 // ListDirFunc - "listDir" function of type listDirFunc returned by listDirFactory() - explained below.
-type ListDirFunc func(bucket, prefixDir, prefixEntry string) (emptyDir bool, entries []string)
+type ListDirFunc func(bucket, prefixDir, prefixEntry string) (emptyDir bool, entries []string, delayIsLeaf bool)
+
+// IsLeafFunc - A function isLeaf of type isLeafFunc is used to detect if an entry is a leaf entry. There are four scenarios
+// where isLeaf should behave differently:
+// 1. FS backend object listing - isLeaf is true if the entry has a trailing "/"
+// 3. Erasure backend object listing - isLeaf is true if the entry is a directory and contains xl.json
```
Comment (poornas) [inline cmd/tree-walk.go]: can scenarios 2 and 4 be documented here also?

**Review by poornas — CHANGES_REQUESTED** [review]

File: cmd/xl-storage.go [hunk 2]
```diff
@@ -750,24 +748,44 @@ func (s *xlStorage) ListDirSplunk(volume, dirPath string, count int) (entries []
...
+	if err != nil {
+		return false
+	}
+
+	_, err = os.Stat(pathJoin(volumeDir, leafPath, xlStorageFormatFile))
+	if err == nil {
+		return true
+	}
+	if os.IsNotExist(err) {
+		if err = s.renameLegacyMetadata(volume, leafPath); err == nil {
```
Comment (vadmeste) [inline cmd/xl-storage.go]: Is this correct ? (why are we renaming anything in isLeaf function) ?

File: cmd/tree-walk.go [hunk 3]
```diff
@@ -51,15 +51,80 @@ func filterMatchingPrefix(entries []string, prefixEntry string) []string {
...
+	return true
+}
+
 // ListDirFunc - "listDir" function of type listDirFunc returned by listDirFactory() - explained below.
-type ListDirFunc func(bucket, prefixDir, prefixEntry string) (emptyDir bool, entries []string)
+type ListDirFunc func(bucket, prefixDir, prefixEntry string) (emptyDir bool, entries []string, delayIsLeaf bool)
+
+// IsLeafFunc - A function isLeaf of type isLeafFunc is used to detect if an entry is a leaf entry. There are four scenarios
+// where isLeaf should behave differently:
+// 1. FS backend object listing - isLeaf is true if the entry has a trailing "/"
```
Comment (vadmeste) [inline cmd/tree-walk.go]: I think you mean if the entry doesn't have a trailing '/'

File: cmd/tree-walk.go [hunk 4]
```diff
@@ -51,15 +51,80 @@ func filterMatchingPrefix(entries []string, prefixEntry string) []string {
...
+// we need to remove this trailing "/" for objects and retain "/" for prefixes before
+// sorting because the trailing "/" can affect the sorting results for certain cases.
+// Ex. lets say entries = ["a-b/", "a/"] and both are objects.
+//     sorting with out trailing "/" = ["a", "a-b"]
+//     sorting with trailing "/"     = ["a-b/", "a/"]
+// Hence if entries[] does not have a case like the above example then isLeaf() check
+// can be delayed till the entry is pushed into the TreeWalkResult channel.
+// delayIsLeafCheck() returns true if isLeaf can be delayed or false if
+// isLeaf should be done in listDir()
+func delayIsLeafCheck(entries []string) bool {
```
Comment (vadmeste) [inline cmd/tree-walk.go:67]: I think 'delaySorting' is a better name here.

**Review by vadmeste — COMMENTED** [review]
Some comments

File: cmd/tree-walk.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/tree-walk.go:67]: No it is delayIsLeaf

File: cmd/xl-storage.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/xl-storage.go]: We should as per the requirement for xl.json to be moved to xl.meta

File: cmd/tree-walk.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/tree-walk.go]: Actually there are no other scenarios @poornas will remove the wording.

**Comment by harshavardhana** [comment]
NOTE: The entire code is actually from previous implementation of Walkers() when they used 
to be at the object layer - the assumptions underneath satisfy all the unit tests we have especially
regarding how the directories v/s objects 

These unit tests have been captured over the years in checking various corner cases which 
comes under S3 compatibility.

File: cmd/xl-storage.go [hunk 5]
```diff
@@ -765,24 +763,49 @@ func (s *xlStorage) ListDirSplunk(volume, dirPath string, count int) (entries []
...
+	}
+	if os.IsNotExist(err) {
+		// We need a fallback code where directory might contain
+		// legacy `xl.json`, in such situation we just rename
+		// and proceed if rename is successful we know that it
+		// is the leaf since `xl.json` was present.
+		if err = s.renameLegacyMetadata(volume, leafPath); err == nil {
+			// if rename was successful, means we did find
+			// old `xl.json` and this is the leaf.
+			return true
```
Comment (klauspost) [inline cmd/xl-storage.go]:
```suggestion
		return  s.renameLegacyMetadata(volume, leafPath) == nil
```
(I think the comment above makes the rest clear)

File: cmd/os-readdir_unix.go [hunk 6]
```diff
@@ -30,10 +30,10 @@ import (
 // refer https://github.com/golang/go/issues/24015
 const blockSize = 8 << 10 // 8192
 
-// By default atleast 1000 entries in single getdents call
+// By default atleast 20 entries in single getdents call
```
Comment (klauspost) [inline cmd/os-readdir_unix.go:33]:
```suggestion
// Allocate at least 20 entries for single getdents call
```

**Review by klauspost — APPROVED** [review]
Looks good. I think I understand it but I assume you vouch for correctness (along with existing tests)

Few minor things.

**Review by poornas — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#6ec22983-8b30-4b06-863a-9324f65f9097) |
<a name="6ec22983-8b30-4b06-863a-9324f65f9097"></a>
### 10346-9dc4b9d/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:32567
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 99b60516271b:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... FAILED in 5 minutes and 39 seconds
{
  "name": "aws-sdk-php",
  "function": "getBucketPolicy ( array $params = [] )",
  "args": {
    "Bucket": "aws-sdk-php-5f4552dc3f1e2"
  },
  "duration": "41529",
  "status": "FAIL",
  "error": "Error executing \"CreateBucket\" on \"http://minio-c2.minio.io:32567/aws-sdk-php-5f4552dc3f1e2\"; AWS HTTP error: Client error: `PUT http://minio-c2.minio.io:32567/aws-sdk-php-5f4552dc3f1e2` resulted in a `404 Not Found` response:\n<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>NoSuchBucket</Code><Message>The specified bucket does not exist</Mes (truncated...)\n NoSuchBucket (client): The specified bucket does not exist - <?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>NoSuchBucket</Code><Message>The specified bucket does not exist</Message><BucketName>aws-sdk-php-5f4552dc3f1e2</BucketName><Resource>/aws-sdk-php-5f4552dc3f1e2</Resource><RequestId>162E95485F609AD5</RequestId><HostId>81966869-52fe-4c4a-8efc-643f19ce8ca0</HostId></Error>"
}
(3/15) Running aws-sdk-ruby tests ... done in 17 seconds
(4/15) Running awscli tests ... done in 2 minutes and 44 seconds
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 3 minutes and 16 seconds
(7/15) Running minio-dotnet tests ... done in 1 minutes and 33 seconds
(8/15) Running minio-go tests ... FAILED in 5 minutes and 9 seconds
{
  "args": {},
  "duration": 1329,
  "error": "The specified multipart upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.",
  "function": "CopyObjectPart(destination, source)",
  "message": "CompleteMultipartUpload call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... done in 6 minutes and 12 seconds
(9/15) Running minio-js tests ... done in 2 minutes and 36 seconds
(10/15) Running minio-py tests ... done in 10 minutes and 11 seconds
(11/15) Running s3cmd tests ... done in 1 minutes and 30 seconds
(12/15) Running s3select tests ... done in 50 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10346-9dc4b9d
